common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+@ThreadSafe
+@Slf4j
+public class ContinuationTokenAsyncIterator<T, U> implements AsyncIterator<U> {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final Queue<U> queue;
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final AtomicReference<T> token;
+    private final Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function;
+
+    public ContinuationTokenAsyncIterator(Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function, T tokenIdentity) {
+        this.function = function;
+        this.token = new AtomicReference<>(tokenIdentity);
+        this.queue = new LinkedBlockingQueue<>();
+    }
+
+    /**
+     * If multiple getNext are called concurrently it will result in multiple remote calls with
+     * same continuation token. This can mean same set of result can get added to the queue multiple times.
+     * To mitigate this, we add result to queue only if the token used in the request and existing value of token are same. 
+     * So if multiple getNext are concurrently called for same continuation token, the result from exactly one of them
```

Comment:
So if there are more than one concurrent getNext calls and our queue was empty, then we had two options --
1. let concurrent function calls with same continuation token be made remotely and only include result from one such call
2. ensure exactly one `function` call is outstanding. 

I went with `1` where i let multiple `function` calls be made and handle the result and put only one result in the queue.

Reply:
this is what the code will look like:

```
    @GuardedBy("lock")
    private CompletableFuture<Void> outstanding;

    @Override
    public CompletableFuture<U> getNext() {
        final T continuationToken;
        synchronized (lock) {
            // if the result is available, return it without making function call
            if (!queue.isEmpty()) {
                return CompletableFuture.completedFuture(queue.poll());
            } else {
                continuationToken = token.get();
                // make the function call if previous outstanding call completed.
                if (outstanding.isDone()) {
                    outstanding = function.apply(continuationToken)
                            .thenAccept(resultPair -> {
                                synchronized (lock) {
                                    if (token.get().equals(continuationToken)) {
                                        log.debug("Received the following collection after calling the function: {} with continuation token: {}",
                                                resultPair.getValue(), resultPair.getKey());
                                        queue.addAll(resultPair.getValue());
                                        token.set(resultPair.getKey());
                                    }
                                }
                            }).whenComplete((x, e) -> {
                                              if (e != null) {
                                                  log.warn("Async iteration failed: ", e);
                                              }
                                          });
                }
            }
        }

        return outstanding.thenCompose(v -> getNext());
    }
```

Reply:
I have made the change. The only thing is we call a method under a lock. And this method is supplied by an external entity to this class. So we dont know how long it will take or what it will do while under the lock before it returns the future. 
I am thinking i can improve that aspect though. I will try and update the code and see if it helps.